### PR TITLE
 Restore after validating the backup location

### PR DESCRIPTION
**Pull request by snigdha-px**
**What this PR does / why we need it**:
Automated PB-3962: Encryption key for backup location is lost if we validate backup location again resulting in restore failure
**Which issue(s) this PR fixes** (optional)
Closes #PA-1284
Log : https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/1824/consoleFull

https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/1825/console
**Special notes for your reviewer**:

### CODE REVIEW

File: tests/backup/backup_portworx_test.go [hunk 1]
```diff
@@ -586,6 +586,19 @@ var _ = Describe("{RestoreEncryptedAndNonEncryptedBackups}", func() {
...
+			err = ValidateBackupLocation(ctx, orgID, backupLocationNames[1], BackupLocation1UID)
+			log.FailOnError(err, "backup location %s validation failed", backupLocationNames[1])
+		})
+		Step("Restore the encrypted backup after validating the backup location", func() {
+			log.InfoD("Restore the encrypted backup after validating the backup location")
+			restoreName := fmt.Sprintf("%s-%s-encrypted", restoreNamePrefix, backupNames[1])
+			restoreNames = append(restoreNames, restoreName)
+			err = CreateRestore(restoreName, backupNames[1], nil, destinationClusterName, orgID, ctx, make(map[string]string))
+			log.FailOnError(err, "%s restore failed", restoreName)
+		})
```
Comment (trenukarya-px) [inline tests/backup/backup_portworx_test.go:601]: Can we add one more backup as well ?

Comment (snigdha-px) [reply tests/backup/backup_portworx_test.go:617]: done

File: tests/backup/backup_portworx_test.go [hunk 2]
```diff
@@ -586,6 +586,35 @@ var _ = Describe("{RestoreEncryptedAndNonEncryptedBackups}", func() {
 			err = CreateRestore(restoreName, backupNames[1], nil, destinationClusterName, orgID, ctx, make(map[string]string))
 			log.FailOnError(err, "%s restore failed", restoreName)
 		})
+		// PB-3962
```
Comment (kshithijiyer-px) [inline tests/backup/backup_portworx_test.go:589]: Can we add some more info here then just the ticket?

Comment (snigdha-px) [reply tests/backup/backup_portworx_test.go]: done

File: tests/backup/backup_helper.go [hunk 3]
```diff
@@ -3410,3 +3410,14 @@ func GetCustomBucketName(provider string, testName string) string {
 	}
 	return customBucket
 }
+
+// ValidateBackupLocation validates the given backup location
+func ValidateBackupLocation(ctx context.Context, orgID string, backupLocationName string, uid string) error {
+	backupLocationInspectRequest := &api.BackupLocationValidateRequest{
```
Comment (kphalgun-px) [inline tests/backup/backup_helper.go:3416]: How about backupLocationValidateRequest?

Comment (snigdha-px) [reply tests/backup/backup_helper.go]: corrected

**Review by kphalgun-px — APPROVED** [review]
LGTM

**Review by kshithijiyer-px — APPROVED** [review]
LGTM

**Review by trenukarya-px — APPROVED** [review]

**Review by ak-px — APPROVED** [review]
LGTM

File: tests/backup/backup_portworx_test.go [hunk 4]
```diff
@@ -586,6 +586,35 @@ var _ = Describe("{RestoreEncryptedAndNonEncryptedBackups}", func() {
...
 			log.FailOnError(err, "%s restore failed", restoreName)
 		})
+		// PB-3962: Taking encrypted backup, validating the backup location and restoring the backup to make sure the encryption key is not lost after backup location validation which will result in restore failure
+		Step("Taking new encrypted backup", func() {
+			log.InfoD("Taking new encrypted backup")
+			for _, namespace := range bkpNamespaces {
+				ctx, err := backup.GetAdminCtxFromSecret()
+				log.FailOnError(err, "Fetching px-central-admin ctx")
+				appContextsToBackup := FilterAppContextsByNamespace(scheduledAppContexts, []string{namespace})
+				encryptedBackupName := fmt.Sprintf("new-%s-%s-%s", "encryption", BackupNamePrefix, namespace)
```
Comment (mkoppal-px) [inline tests/backup/backup_portworx_test.go:596]: nit: Can you make it `encrypted` ?

Comment (snigdha-px) [reply tests/backup/backup_portworx_test.go:596]: done

File: tests/backup/backup_portworx_test.go [hunk 5]
```diff
@@ -586,6 +586,35 @@ var _ = Describe("{RestoreEncryptedAndNonEncryptedBackups}", func() {
...
+			err = ValidateBackupLocation(ctx, orgID, backupLocationNames[1], BackupLocation1UID)
+			log.FailOnError(err, "backup location %s validation failed", backupLocationNames[1])
+		})
+
+		Step("Restore the encrypted backups after validating the encrypted backup location", func() {
+			log.InfoD("Restore the encrypted backups after validating the encrypted backup location")
+			for i := 1; i < len(backupNames); i++ {
+				restoreName := fmt.Sprintf("%s-%s-encrypted", restoreNamePrefix, backupNames[i])
+				restoreNames = append(restoreNames, restoreName)
+				err = CreateRestore(restoreName, backupNames[i], nil, destinationClusterName, orgID, ctx, make(map[string]string))
```
[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/backup_portworx_test.go:614]: done

[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/backup_portworx_test.go]: Pass log after making the changes: https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/1867/consoleFull

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Pull request merged**